### ISSUE TITLE
Remove host permission for easier to audit

### ISSUE THREAD

**Issue by newbe36524** [mingyaulee/Blazor.BrowserExtension#4]
## Background

When trying to upload a extension to the chrome extensions store, it can take longer to review the extension if it requires "host permission".
This is explained in the review prompt page when uploading.
Therefore, we would like to explore a solution to remove the need for "host permission", i.e. remove the need for `"*:///*/*"` from mainfest.json.

## Proposal

It is recommended that developers set all Razor page URLs to end with index.html.

For example.

- /manager/index.html
- /popup/index.html
- /background/index.html

With this setting, users can not only use NavigationManager to jump directly to the page. It is possible to open the interface with a browser, for example /manager/index.html, without getting 404 when refreshing with F5.

To support index.html generation in each path.
You need to automatically create a folder with index.html in the above location and copy index.html to it during building project.

2. Add a helper class to help developers quickly splice the above special links, for example, to achieve the following effect.

```cs
await WebExtension.Tabs.Create(new
{
    url = "/index.html?path=/Manager/index.html"
});
```

It is possible to use.

```cs
await WebExtension.Tabs.Create(new
{
    url = WebExtension.GetPageUrl("/Manager")
});

**Comment by mingyaulee** [comment]
This is a good idea. I will get to work on it tonight, unless if you want to implement this.

**Comment by newbe36524** [comment]
Sorry, I got no idea about msbuild task.I think you are the best person for the job.

**Comment by newbe36524** [comment]
I'm a little curious if we can also remove 'webRequest' and 'webRequestBlocking' if we implement them in this way

**Comment by mingyaulee** [comment]
I had some thought about this and I don't think defining all Razor page URLs to end with index.html will help with removing the host permissions.

The reason being that chrome extensions only serve real physical files, where the structure of the wwwroot is
- wwwroot
    - index.html

To be able to remove host permission, webRequest and webRequestBlocking, the structure of the wwwroot needs to be 
- wwwroot
    - manager
        - index.html
    - popup
        - index.html
    - background
        - index.html
where all the index.html files are exactly the same. This requirement cannot be easily done from the build process, thus there is a need for the background script to be actively intercepting the requests and redirecting chrome to the wwwroot/index.html file.

With that said, I think it is probably good to remove the validation for the manifest to specify webRequest and webRequestBlocking, and the background script can check for permission during runtime.

**Comment by newbe36524** [comment]
1. Is that possible to detect all .razor file which contains a `@page "/**/index.html"` line in thet file. And try to copy them?
2. Yes, I think a need to build property to skip mainfest validation.

I build a bookmark manager extension and in that project, I just create that index.html files manually.
Now, there is a error if I remove 'webRequest' and 'webRequestBlocking'

**Comment by mingyaulee** [comment]
To answer your questions:
1. Yes it is possible to do this during build
2. I have removed the validation and published a new release

**Comment by mingyaulee** [comment]
To answer your questions:
1. Yes it is possible to do this during build
2. I have removed the validation and published a new release

**Comment by newbe36524** [comment]
Many thanks, it works.

**Comment by mingyaulee** [comment]
> 1. Is that possible to detect all .razor file which contains a @page "/**/index.html" line in thet file. And try to copy them?

@newbe36524 this idea has been implemented in version [v0.4.2](https://github.com/mingyaulee/Blazor.BrowserExtension/releases/tag/v0.4.2) if you are interested to try it out.

Also, out of topic, I have implemented an HttpClient that supports extensions protocol which I think you'll find useful. You can check it out in the release notes of [v0.4.0](https://github.com/mingyaulee/Blazor.BrowserExtension/releases/tag/v0.4.0).

**Comment by newbe36524** [comment]
@mingyaulee Thanks to your hard work  it works well !
